hazelcast/src/main/java/com/hazelcast/config/MapConfig.java

```diff
@@ -674,15 +679,66 @@ public boolean isNearCacheEnabled() {
         return nearCacheConfig != null;
     }
 
+    /**
+     *
+     * @return
+     * @deprecated user {@link #getQueryCacheConfigs()} instead.
+     */
     public boolean isOptimizeQueries() {
-        return optimizeQueries;
+        return cacheDeserializedValues == CacheDeserializedValues.ALWAYS;
     }
 
+    /**
+     *
+     * Enable de-serialized value caching when evaluating predicates.
+     *
+     * @param optimizeQueries
+     * @return this {@code MapConfig} instance.
+     * @deprecated use {@link #setCacheDeserializedValues(CacheDeserializedValues)} instead
+     */
     public MapConfig setOptimizeQueries(boolean optimizeQueries) {
-        this.optimizeQueries = optimizeQueries;
+        if (optimizeQueries) {
+            if (cacheDeserializedValues == DEFAULT_CACHED_DESERIALIZED_VALUES) {
+                cacheDeserializedValues = CacheDeserializedValues.ALWAYS;
+            } else if (cacheDeserializedValues == CacheDeserializedValues.NEVER) {
+                throw new ConfigurationException("Deprecated option 'optimize-queries' is set to true, "
+                        + "but 'cacheDeserializedValues' is set to NEVER. These are conflicting options. Please "
+                        + "remove the `optimize-queries'");
+            }
+        } else {
+            if (cacheDeserializedValues == CacheDeserializedValues.ALWAYS) {
+                throw new ConfigurationException("Deprecated option 'optimize-queries' is set to false, "
+                        + "but 'cacheDeserializedValues' is set to ALWAYS. These are conflicting options. Please "
+                        + "remove the `optimize-queries'");
+            }
+        }
         return this;
     }
 
+    /**
+     * Configure value de-serialized value caching.
+     * Default: {@link CacheDeserializedValues#INDEX_ONLY}
+     *
+     * @param cacheDeserializedValues
+     * @return this {@code MapConfig} instance.
+     * @since 3.6
+     * @see {@link CacheDeserializedValues}
+     */
+    public MapConfig setCacheDeserializedValues(CacheDeserializedValues cacheDeserializedValues) {
```

Comment:
Ca you also add to javaDoc here that what will be the behaviour when InMemoryFormat is OBJECT? Or is this only valid for InMemoryFormat#BINARY ?